docs/architecture.md

```diff
@@ -221,10 +222,20 @@ The query frontend supports caching query results and reuses them on subsequent
 
 The **ruler** is an **optional service** executing PromQL queries for recording rules and alerts. The ruler requires a database storing the recording rules and alerts for each tenant.
 
-Ruler can be scaled horizontally.
+Ruler are **stateless** and can be scaled horizontally.
 
 ### Alertmanager
 
 The **alertmanager** is an **optional service** responsible for accepting alert notifications from the [ruler](#ruler), deduplicating and grouping them, and routing them to the correct notification channel, such as email, PagerDuty or OpsGenie.
 
 The Cortex alertmanager is built on top of the [Prometheus Alertmanager](https://prometheus.io/docs/alerting/alertmanager/), adding multi-tenancy support. Like the [ruler](#ruler), the alertmanager requires a database storing the per-tenant configuration.
+
+Alertmanager are **stateful**.
```

Comment:
Could you elaborate why you consider the Alertmanager stateful? It's very likely I'm missing something obvious.

@jtlisi Do you think it's correct stating that Alertmanager is stateful while Ruler is not?

Reply:
The Peometheus Alertmanager need to store alert state and silence rule.
I think Cortex one also need it. And cortex alertmanager have this config:
```
# Base path for data storage.
# CLI flag: -alertmanager.storage.path
[datadir: <string> | default = "data/"]
```

Reply:
Yes, true.

I'm not an expert of it, but if clustering is used it's somewhat half stateful, in the sense that you don't necessarily need data persistency cause data is replicated across alertmanager instances (and you end up with state/silence loss only if you lose all instances at once).

@jtlisi Could you share more details to better phrase this stateful thing in the doc, please?

Reply:
Both of these services can be considered stateful to different degrees. The Alertmanager persists information about silences and active alerts to its disk. If all of the alertmanager nodes failed simultaneously there would be a loss of data.

The ruler is stateful as well. Running rules internally have state, as well as the ring the rulers initiate. However, if the rulers all fail and restart, Prometheus alert rules have a feature where an alert is restored and returned to a firing state if it would have been active in its `for` period. However, there would be gaps in the series generated by the recording rules.

I don't think the term "stateful" is particularly helpful in this context. It might be better to explain how the Alertmanager persists it's the state of alerts and active silences to a local disk and that it is recommended they are run using a stateful set.